mono/metadata/class.c

```diff
@@ -2202,18 +2204,40 @@ mono_class_layout_fields (MonoClass *klass, int base_instance_size, int packing_
 			mono_class_set_type_load_failure (klass, "Value type instance size (%d) cannot be zero, negative, or bigger than 1Mb", klass->instance_size);
 	}
 
+	// Weak field support
+	//
+	// FIXME:
+	// - generic instances
+	// - Disallow on structs/static fields/nonref fields
+	gboolean has_weak_fields = FALSE;
+
+	if (mono_class_has_static_metadata (klass)) {
+		for (MonoClass *p = klass; p != NULL; p = p->parent) {
+			gpointer iter = NULL;
+			guint32 first_field_idx = mono_class_get_first_field_idx (p);
+
+			while ((field = mono_class_get_fields (p, &iter))) {
+				guint32 field_idx = first_field_idx + (field - p->fields);
+				if (MONO_TYPE_IS_REFERENCE (field->type) && mono_assembly_is_weak_field (p->image, field_idx + 1)) {
+					has_weak_fields = TRUE;
+					mono_trace_message (MONO_TRACE_TYPE, "Field %s:%s at offset %x is weak.", field->parent->name, field->name, field->offset);
```

Comment:
Why does it have to keep scanning?

Reply:
sure but we are looking only for the first one, right?

Reply:
We are also printing out the trace message. Since these fields are not common, I don't think continuing the loop is a problem.

Reply:
There's still the issue that we're iterating over the parents multiple times.
This info should be cached in class and not recomputed multiple times for the same class.

Plus, this breaks SRE if it has a parent with a Weak field.